### ISSUE TITLE
`--nocapture` doesn't follow common CLI conventions, making it a stumbling block to people debugging failures

### ISSUE THREAD

**Issue by epage** [rust-lang/rust#133073]
By convention, users would expect to type in `--no-capture`.  The fact that the argument is `--nocapture` trips people up, especially as they have to wait for their test to compile before they see the failure.  Without spelling suggestions, they need to then consult the help to then remember its without the middle `-`.  Unless someone is doing this all the time to build up muscle memory to counteract intuition, this will trip people up each time.

See also
- #24451
- https://hachyderm.io/@wezm@mastodon.decentralised.social/113485301109871075
- https://rust-lang.zulipchat.com/#narrow/channel/404371-t-testing-devex/topic/--nocapture/near/482538793

**Comment by epage** [comment]
```
Usage: /home/epage/src/personal/cargo/target/debug/deps/cargo-ef836591b671bcbc [OPTIONS] [FILTERS...]

Options:
        --include-ignored
                        Run ignored and not ignored tests
        --ignored       Run only ignored tests
        --force-run-in-process
                        Forces tests to run in-process when panic=abort
        --exclude-should-panic
                        Excludes tests marked as should_panic
        --test          Run tests and not benchmarks
        --bench         Run benchmarks instead of tests
        --list          List all tests and benchmarks
    -h, --help          Display this message
        --logfile PATH  Write logs to the specified file
        --nocapture     don't capture stdout/stderr of each task, allow
                        printing directly
        --test-threads n_threads
                        Number of threads used for running tests in parallel
        --skip FILTER   Skip tests whose names contain FILTER (this flag can
                        be used multiple times)
    -q, --quiet         Display one character per test instead of one line.
                        Alias to --format=terse
        --exact         Exactly match filters rather than by substring
        --color auto|always|never
                        Configure coloring of output:
                        auto = colorize if stdout is a tty and tests are run
                        on serially (default);
                        always = always colorize output;
                        never = never colorize output;
        --format pretty|terse|json|junit
                        Configure formatting of output:
                        pretty = Print verbose output;
                        terse = Display one character per test;
                        json = Output a json document;
                        junit = Output a JUnit document
        --show-output   Show captured stdout of successful tests
    -Z unstable-options Enable nightly-only flags:
                        unstable-options = Allow use of experimental features
        --report-time   Show execution time of each test.
                        Threshold values for colorized output can be
                        configured via
                        `RUST_TEST_TIME_UNIT`, `RUST_TEST_TIME_INTEGRATION`
                        and
                        `RUST_TEST_TIME_DOCTEST` environment variables.
                        Expected format of environment variable is
                        `VARIABLE=WARN_TIME,CRITICAL_TIME`.
                        Durations must be specified in milliseconds, e.g.
                        `500,2000` means that the warn time
                        is 0.5 seconds, and the critical time is 2 seconds.
                        Not available for --format=terse
        --ensure-time   Treat excess of the test execution time limit as
                        error.
                        Threshold values for this option can be configured via
                        `RUST_TEST_TIME_UNIT`, `RUST_TEST_TIME_INTEGRATION`
                        and
                        `RUST_TEST_TIME_DOCTEST` environment variables.
                        Expected format of environment variable is
                        `VARIABLE=WARN_TIME,CRITICAL_TIME`.
                        `CRITICAL_TIME` here means the limit that should not
                        be exceeded by test.
        --shuffle       Run tests in random order
        --shuffle-seed SEED
                        Run tests in random order; seed the random number
                        generator with SEED

```
For casing, the two that are questionable are
- `--logfile` which T-testing-devex has an FCP for deprecating, see rust-lang/testing-devex-team#9
- `--nocapture` (I just put the middle `-` in...)

For consistency with the rest of Rust tools
- `--test-threads`: Cargo commands refer to this as `-j --jobs`

For consistency with [pytest](https://docs.pytest.org/en/stable/reference/reference.html#command-line-flags)
- `--no-capture` is instead `--capture=no` with a short flag of `-s`
  - Other values `fd|sys|no|tee-sys`

For inspiration from [cargo-nextest](https://nexte.st/)
- `--ignored`, `--include-ignored`: `--run-ignored <default|only|all>`
- `--test-threads`: `-j --thread-threads`

If people have other test runners they want to compare, you are welcome to add them!  I'm scoping this to the existing, stable features of libtest.

**Comment by jieyouxu** [comment]
@epage would you like to open a tracking issue for the CLI audit, and have *this* issue focused on `--nocapture` specifically?

**Comment by epage** [comment]
My hope is that we can close out the discussion quickly enough that we don't need to split it up.  We'll likely need to do an FCP anyways for `--nocapture`, so might as well look at it holistically.

**Comment by jieyouxu** [comment]
Understood. I have a datapoint about the behavior of `--skip` and general test filtering (not sure if this fits with this issue):

- `--skip` and `--exact` is not very intuitive and can have very surprising behavior: it feels like the fuzzy vs exact match should be flipped, and have `--fuzzy` instead of `--exact`. e.g. we ran into this in https://github.com/rust-lang/rust/pull/132979. The filter will match by substring.
> Without the --exact flag, using --skip tests/rustdoc can unintentionally skip other tests that match as substrings such as rustdoc-gui, rustdoc-js, etc.
- The same is true for the general test filtering, e.g. `test -- my/test/filter` uses AFAIK a substring match which can be very surprising.

**Comment by epage** [comment]
Hmm, I see why you brought up a tracking issue.  My focus on this is purely in naming and not in improving the semantics.  My expectation is that we'll be moving filtering into Cargo and re-evaluate the interface then.  Moving more logic into Cargo will be unblocked by the json format being stabilized (so Cargo communicates through tests through json) and this will reduce the burden of what custom test harnesses need to implement.

**Comment by jieyouxu** [comment]
Makes sense, I'll mark my comment as off-topic for this issue to keep this issue focused.

**Comment by epage** [comment]
We talked about this in today's T-testing-devex meeting

- `--nocapture`: we were on board with changing this to `--no-capture` through deprecation/addition
- `--test-threads`: we are deferring this to see how things evolve for thread control within the compiler
- We didn't see anything else worth changing at this time with the assumption that most of the UX responsibility will be shifting to tools calling libtest, like `cargo test`.

Let's see if this works

@rfcbot fcp merge

**Comment by djc** [comment]
@epage thanks for pushing to clean this up!!

**Comment by epage** [comment]
Something we had overlooked was that `RUST_TEST_NOCAPTURE` exists which checks for a non-zero value to disable capturing (including empty).

Options
- Defer this
- Add `RUST_TEST_NO_CAPTURE`
- Switch to positive logic with `RUST_TEST_CAPTURE`
  - Maybe re-evaluate the true/false condition
- Deprecate it

Thoughts?

**Comment by djc** [comment]
Adding `RUST_TEST_NO_CAPTURE` would make sense to me. Switching to positive logic (while I usually prefer it) feels like a mismatch with the CLI, so maybe it doesn't make sense here.

**Comment by weihanglo** [comment]
It seems reasonable to do the same deprecation to env vars.

**Comment by epage** [comment]
Thinking more on this, I'm leaning towards "defer".

`--no-capture` is a nice quality of life improvement that I'd rather not block on "bigger questions".  For me, the big question I'm starting to wonder about is if we should deprecate the env variables.   My suspicion is that they are a low-effort config system that is meant to make up for a lacking in the common test runner (`cargo test`).  In that case, we should instead focus on the common test runner.  These could also cause problems as Cargo tries to coordinate the behavior of test binaries.  If `cargo test` is trying to read json output and the user bypasses cargo to tell the test to report other content, that could become problematic.
